org.mwc.debrief.track_shift/src/org/mwc/debrief/track_shift/controls/ZoneChart.java

```diff
@@ -78,6 +78,54 @@ private CustomChartComposite(final Composite parent, final JFreeChart chart)
 
     private Stack<Rectangle> zoomHistory = new Stack<Rectangle>();
 
+    void fitToData()
+    {
+      final List<Rectangle> zoomRedoHistory = new ArrayList<>(zoomHistory);
+
+      AbstractOperation addOp = new AbstractOperation("Show all data")
+      {
+
+        @Override
+        public IStatus execute(IProgressMonitor monitor, IAdaptable info)
+            throws ExecutionException
+        {
+
+          zoomHistory.clear();
+
+          CustomChartComposite.super.restoreAutoBounds();
+          return Status.OK_STATUS;
+        }
+
+        @Override
+        public IStatus redo(IProgressMonitor monitor, IAdaptable info)
+            throws ExecutionException
+        {
+          zoomHistory.clear();
+
+          CustomChartComposite.super.restoreAutoBounds();
+          return Status.OK_STATUS;
+        }
+
+        @Override
+        public IStatus undo(IProgressMonitor monitor, IAdaptable info)
+            throws ExecutionException
+        {
+
+          CustomChartComposite.super.restoreAutoBounds();
+          zoomHistory.addAll(zoomRedoHistory);
+
+          for (Rectangle h : zoomHistory)
+          {
```

Comment:
You've got me stuck here.  I can't see why the **undo** needs us to loop through a set of zoom operations.  I also can't see understand the roles of *zoomRedoHistory* and  *zoomHistory* objects.  We seem to create *zoomRedoHistory* using the contents of *zoomHistory*. But, we also fill *zoomHistory* with the contents of *zoomRedoHistory*.

I would think that an easier strategy would be to remember the existing rectangle when we do *execute* before we fit to win. Then we restore that rectangle in an *undo*.   Am I missing something?